File: src/backend/routes/night-queue.ts
import { FastifyInstance } from 'fastify';
import { loadConfig, resolveAssistantKey, resolveEnvVars } from '../config.js';
import { createHermesClient, type HermesFetch } from '../hermes/client.js';
import type { NexusConfig } from '@nexus/shared';

interface NightQueueRoutesOptions {
  fetchImpl?: HermesFetch;
}

// Read-only proxy over the partner adapter's night-queue board (`GET
// /v1/night-queue`, baker-internal#111) — what the overnight runner (#55) did,
// what is queued for tonight, and which of its PRs are still open for Keith.
// No controls: the queue is filled by minting a label in daytime discussion,
// never from a dashboard.
export function createNightQueueRoutes(
  load: () => NexusConfig = loadConfig,
  options: NightQueueRoutesOptions = {},
) {
  return async function registerNightQueueRoutes(fastify: FastifyInstance) {
    const client = () => {
      const config = load();
      const url = resolveEnvVars(config.assistant.url || '').trim();
      const key = resolveAssistantKey(config);
      if (!url || !key) return undefined;
      return createHermesClient({ url, key, fetchImpl: options.fetchImpl });
    };

    // Fail-soft like /api/routines: an unconfigured or unreachable adapter
    // yields empty sections plus a reason, never a 5xx. `available: false` is
    // the adapter's own "no night has run yet" and is passed through as data —
    // the card distinguishes "nothing has happened" from "I cannot tell".
    fastify.get('/api/night-queue', async (request) => {
      const { nights } = request.query as { nights?: string };
      const hermes = client();
      const empty = { nights: [], queue: [], open_prs: [] };
      if (!hermes) return { configured: false, available: false, ...empty };
      try {
        const parsed = Number(nights);
        const report = (await hermes.listNightQueue(
          Number.isFinite(parsed) && parsed > 0 ? parsed : undefined,
        )) as Record<string, unknown>;
        return { configured: true, available: false, ...empty, ...report };
      } catch (err: any) {
        return {
          configured: true,
          available: false,
          ...empty,
          error: err?.message || 'Night queue fetch failed.',
        };
      }
    });

    fastify.get('/api/night-queue/nights/:nightId', async (request, reply) => {
      const { nightId } = request.params as { nightId: string };
      const hermes = client();
      if (!hermes) {
        reply.code(400);
        return { error: 'Assistant URL and key must be configured in Settings.' };
      }
      try {
        return await hermes.getNight(nightId);
      } catch (err: any) {
        const message: string = err?.message || 'Night fetch failed.';
        reply.code(/not found/i.test(message) ? 404 : 502);
        return { error: message };
      }
    });
  };
}

export const registerNightQueueRoutes = createNightQueueRoutes();

File: ios/NexusCore/Sources/NexusCore/Models/NightQueue.swift
import Foundation

/// Night-queue board (`GET /api/night-queue`, proxied from the assistant
/// adapter's `/v1/night-queue` — baker-internal#111). Mirror of
/// `NightQueueResponse` in `src/frontend/src/api.ts`. Wire keys are
/// snake_case, so the shared `.nexusREST` decoder applies.
///
/// Read-only surface: the queue is filled by minting a `night-queue` label in
/// daytime discussion, never from a dashboard.
public struct NightQueueResponse: Decodable, Sendable {
    /// False when the backend has no assistant URL/key configured.
    public let configured: Bool?
    /// False until the overnight runner has written its first night. Distinct
    /// from `error`: "nothing has happened yet" is not "I cannot tell".
    public let available: Bool?
    public let nights: [Night]
    /// What the scheduled launchd job last did. Absent on older adapters.
    public let lastAttempt: NightAttempt?
    public let queue: [QueuedIssue]
    public let openPrs: [NightQueuePR]
    /// Per-section reasons. `*Stale` means the list below is the last good
    /// answer, not a current one.
    public let queueError: String?
    public let queueStale: Bool?
    public let openPrsError: String?
    public let openPrsStale: Bool?
    public let generatedAt: Int?
    /// Set when the backend could not reach the adapter at all.
    public let error: String?
}

/// What the scheduled launchd job last did, from the wrapper's status file.
///
/// The ledger cannot answer "did the job even run?": a night that dies before
/// the runner opens it writes no row, so between 2026-08-29 and 09-01 the
/// runner crashed on an import four nights running and the board showed a
/// four-day-old night as if all were well. This is the only evidence such an
/// attempt happened.
public struct NightAttempt: Decodable, Sendable {
    public let started: Int?
    public let ended: Int?
    public let rc: Int?
    public let timedOut: Bool?
    /// "status" (the wrapper's JSON) or "stamp" (a legacy `.last` file).
    public let source: String?
    /// `rc == 0`. A legacy stamp carries no exit code and reports false —
    /// unknown is not success.
    public let ok: Bool
    /// Whether a ledger night covers this attempt. Independent of `ok`: a run
    /// that exited 0 and wrote no night is still unaccounted for. `nil` when
    /// there is no ledger to check, which is not the same as "went missing".
    public let recorded: Bool?

    /// The one state worth interrupting the reader for.
    public var isUnaccountedFor: Bool { recorded == false }
}

/// `quiet` — nothing was labelled. The normal night, and not a failure.
public enum NightOutcome: String, Decodable, Sendable {
    case worked, quiet, running, unknown

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = NightOutcome(rawValue: raw) ?? .unknown
    }
}

/// What the RUNNER observed, never the coder model's claim about its own
/// homework. Decoded as an optional on `NightRun`, where a JSON `null` means
/// the ledger row predates the column — unknown, which is not `notRun`.
public enum NightTests: String, Decodable, Sendable {
    case passed
    case failed
    case notRun = "not_run"
    case unknown

    public init(from decoder: Decoder) throws {
        let raw = try decoder.singleValueContainer().decode(String.self)
        self = NightTests(rawValue: raw) ?? .unknown
    }
}

public struct Night: Decodable, Sendable, Identifiable {
    public let id: String
    public let startedAt: String?
    public let startedTs: Int?
    public let endedAt: String?
    public let endedTs: Int?
    /// drained | window_closed | token_budget | max_issues | fatal
    public let stopReason: String?
    public let issuesPlanned: Int
    public let issuesAttempted: Int
    public let tokensUsed: Int
    public let outcome: NightOutcome
    /// Rollups the adapter computes so web and iOS cannot disagree.
    public let prsOpened: Int
    /// PRs opened without a green test run from the runner itself.
    public let unvalidated: Int
    public let failures: Int
    public let runs: [NightRun]
    /// Only present on the night-detail endpoint.
    public let plan: NightPlan?
}

public struct NightRun: Decodable, Sendable, Identifiable {
    public let id: String
    public let repo: String
    public let issueNumber: Int
    public let branch: String?
    public let startedAt: String?
    public let startedTs: Int?
    public let endedAt: String?
    public let endedTs: Int?
    /// pr_opened | parked | no_changes | timeout | failed
    public let status: String?
    public let rounds: Int
    /// approve | arbitrated_ship | arbitrated_park | unreviewed
    public let verdict: String?
    public let prUrl: String?
    public let tokensUsed: Int
    public let error: String?
    public let summary: String
    /// `nil` = the ledger row predates the column (unknown), NOT `notRun`.
    public let tests: NightTests?
    public let issueUrl: String?

    /// A PR the runner could not prove. The board must never let the PR link
    /// be the loudest thing on such a row.
    public var isUnvalidatedPR: Bool { status == "pr_opened" && tests != .passed }
}

/// What the planner decided. Deliberately carries no issue body or comments —
/// the adapter strips them, since untrusted GitHub prose has no business being
/// re-served to a dashboard.
public struct NightPlan: Decodable, Sendable {
    public let selected: [Selected]
    public let parked: [Parked]
    public let excluded: [Excluded]

    public struct Selected: Decodable, Sendable, Identifiable {
        public var id: String { "\(repo ?? "?")#\(number ?? 0)" }
        public let repo: String?
        public let number: Int?
        public let title: String?
        public let model: String?
        public let budgetTokens: Int?
        public let rationale: String?
    }

    public struct Parked: Decodable, Sendable, Identifiable {
        public var id: String { "\(repo ?? "?")#\(number ?? 0)" }
        public let repo: String?
        public let number: Int?
        public let reason: String?
    }

    public struct Excluded: Decodable, Sendable, Identifiable {
        public var id: String { "\(repo ?? "?")#\(number ?? 0)" }
        public let repo: String?
        public let number: Int?
        public let title: String?
    }
}

/// An open `night-queue`-labelled issue: what tonight's 01:00 run will consider.
public struct QueuedIssue: Decodable, Sendable, Identifiable {
    public var id: String { "\(repo)#\(number)" }
    public let repo: String
    public let number: Int
    public let title: String
    public let url: String
    public let updatedAt: String?
    public let updatedTs: Int?
    /// baker-internal and nexus: labelled, but standing policy means the
    /// unattended agent never touches its own runtime or this surface.
    public let excluded: Bool
    /// The spec the planner will judge — the newest human comment, or the
    /// issue body. Never the runner's own bookkeeping chatter.
    public let readiness: String?
    /// "comment" or "body".
    public let readinessSource: String?
}

/// An open `night-queue-output` PR: work waiting on Keith.
public struct NightQueuePR: Decodable, Sendable, Identifiable {
    public var id: String { "\(repo)#\(number)" }
    public let repo: String
    public let number: Int
    public let title: String
    public let url: String
    public let createdAt: String?
    public let createdTs: Int?
    public let isDraft: Bool
}
